cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -173,137 +151,133 @@ int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
-            nextEventCycles = min_u32(nextEventCycles, wave->nextTimeHighCycles);
-          } else {
+            wave->nextServiceCycle += wave->nextTimeHighCycles;
+            if (nextEventCycles > wave->nextTimeHighCycles)
+              nextEventCycle = wave->nextServiceCycle;
+          } else if ((waveformsState & mask) || !wave->nextTimeHighCycles) {
+            waveformsState &= ~mask;
             if (i == 16) {
               GP16O &= ~1; // GPIO16 write slow as it's RMW
-            } else {
+            }
+            else {
               ClearGPIO(mask);
             }
-            wave->nextServiceCycle = now + wave->nextTimeLowCycles;
-            nextEventCycles = min_u32(nextEventCycles, wave->nextTimeLowCycles);
+            wave->nextServiceCycle += wave->nextTimeLowCycles;
+            if (nextEventCycles > wave->nextTimeLowCycles)
+              nextEventCycle = wave->nextServiceCycle;
           }
-        } else {
-          uint32_t deltaCycles = wave->nextServiceCycle - now;
-          nextEventCycles = min_u32(nextEventCycles, deltaCycles);
         }
+
+        // Disable any waveforms that are done
+        if ((ExpiryState::ON == wave->hasExpiry) && expiryToGo <= 0) {
+          // Done, remove!
+          waveformsEnabled &= ~mask;
+        }
+        now = ESP.getCycleCount();
       }
 
       // Exit the loop if we've hit the fixed runtime limit or the next event is known to be after that timeout would occur
-      uint32_t now = GetCycleCountIRQ();
-      int32_t cycleDeltaNextEvent = timeoutCycle - (now + nextEventCycles);
-      int32_t cyclesLeftTimeout = timeoutCycle - now;
-      done = (cycleDeltaNextEvent < 0) || (cyclesLeftTimeout < 0);
+      done = (now - isrStart >= isrTimeoutCycles) || (nextEventCycle - isrStart  >= isrTimeoutCycles);
     } while (!done);
-  } // if (waveformEnabled)
+  } // if (waveformsEnabled)
 
-  if (timer1CB) {
-    nextEventCycles = min_u32(nextEventCycles, timer1CB());
+  int32_t nextEventCycles;
+  if (!timer1CB) {
+    nextEventCycles = nextEventCycle - ESP.getCycleCount();
   }
-
-  if (nextEventCycles < microsecondsToClockCycles(10)) {
-    nextEventCycles = microsecondsToClockCycles(10);
+  else {
+    int32_t callbackCycles = microsecondsToClockCycles(timer1CB());
+    nextEventCycles = nextEventCycle - ESP.getCycleCount();
+    if (nextEventCycles > callbackCycles)
+      nextEventCycles = callbackCycles;
   }
-  nextEventCycles -= DELTAIRQ;
-
-  // Do it here instead of global function to save time and because we know it's edge-IRQ
-#if F_CPU == 160000000
-  T1L = nextEventCycles >> 1; // Already know we're in range by MAXIRQUS
-#else
-  T1L = nextEventCycles; // Already know we're in range by MAXIRQUS
-#endif
-  TEIE |= TEIE1; // Edge int enable
+
+  if (nextEventCycles < microsecondsToClockCycles(2))
+    nextEventCycles = microsecondsToClockCycles(2);
+
+  if (clockCyclesPerMicrosecond() == 160)
+    timer1_write(nextEventCycles / 2);
```

Comment:
Calling t1_write() will increase code size and runtime thru each iteration.  I appreciate the inelegance of twiddling registers directly here, but it was that way for a reason.

Reply:
Mhm, I though after checking the code that it wasn't worth the distraction, boiling down to basically the same. I haven't checked the exact assembly, admittedly - are you sure that what you remember still actually holds?